include/circt/Support/BackedgeBuilder.h

```diff
@@ -27,7 +30,25 @@ class Operation;
 
 namespace circt {
 
-class Backedge;
+/// `Backedge` is a wrapper class around a `Value`. When assigned another
+/// `Value`, it replaces all uses of itself with the new `Value` then become a
+/// wrapper around the new `Value`.
+class Backedge {
+public:
+  Backedge(mlir::Operation *op) : value(op->getResult(0)){};
+
+  operator mlir::Value() { return value; }
+  void setValue(mlir::Value newValue) {
+    assert(value.getType() == newValue.getType());
+    assert(!set && "backedge already set to a value!");
+    value.replaceAllUsesWith(newValue);
+    set = true;
+  }
+
+private:
+  mlir::Value value;
+  bool set = false;
```

Comment:
I know this is existing behavior, but since this PR brought up thread safety, does this flag need to be a std::atomic or something like that?

Reply:
I don't think so. BackedgeBuilder was never intended to be thread-safe. Its purpose is to be used in a single function and destructed at the end of the function.